db/util.go

```diff
@@ -1,82 +1,46 @@
 package db
 
-import "bytes"
+import (
+	"bytes"
+	"strings"
+)
 
-// A wrapper around itr that tries to keep the iterator
-// within the bounds as defined by `prefix`
-type prefixIterator struct {
-	itr     Iterator
-	prefix  []byte
-	invalid bool
-}
-
-func (pi *prefixIterator) Seek(key []byte) {
-	if !bytes.HasPrefix(key, pi.prefix) {
-		pi.invalid = true
-		return
-	}
-	pi.itr.Seek(key)
-	pi.checkInvalid()
-}
-
-func (pi *prefixIterator) checkInvalid() {
-	if !pi.itr.Valid() {
-		pi.invalid = true
-	}
-}
-
-func (pi *prefixIterator) Valid() bool {
-	if pi.invalid {
-		return false
-	}
-	key := pi.itr.Key()
-	ok := bytes.HasPrefix(key, pi.prefix)
-	if !ok {
-		pi.invalid = true
-		return false
-	}
-	return true
-}
-
-func (pi *prefixIterator) Next() {
-	if pi.invalid {
-		panic("prefixIterator Next() called when invalid")
+func IteratePrefix(db DB, prefix []byte) Iterator {
+	var start, end []byte
+	if len(prefix) == 0 {
+		start = BeginningKey()
+		end = EndingKey()
+	} else {
+		start = cp(prefix)
+		end = cpIncr(prefix)
 	}
-	pi.itr.Next()
-	pi.checkInvalid()
+	return db.Iterator(start, end)
 }
 
-func (pi *prefixIterator) Prev() {
-	if pi.invalid {
-		panic("prefixIterator Prev() called when invalid")
-	}
-	pi.itr.Prev()
-	pi.checkInvalid()
-}
+//----------------------------------------
 
-func (pi *prefixIterator) Key() []byte {
-	if pi.invalid {
-		panic("prefixIterator Key() called when invalid")
-	}
-	return pi.itr.Key()
+func cp(bz []byte) (ret []byte) {
+	ret = make([]byte, len(bz))
+	copy(ret, bz)
+	return ret
 }
 
-func (pi *prefixIterator) Value() []byte {
-	if pi.invalid {
-		panic("prefixIterator Value() called when invalid")
+// CONTRACT: len(bz) > 0
+func cpIncr(bz []byte) (ret []byte) {
+	ret = cp(bz)
+	for i := len(bz) - 1; i >= 0; i-- {
+		if ret[i] < byte(0xFF) {
+			ret[i] += 1
+			return
+		} else {
+			ret[i] = byte(0x00)
+		}
 	}
-	return pi.itr.Value()
+	return EndingKey()
 }
 
-func (pi *prefixIterator) Close()          { pi.itr.Close() }
-func (pi *prefixIterator) GetError() error { return pi.itr.GetError() }
-
-func IteratePrefix(db DB, prefix []byte) Iterator {
-	itr := db.Iterator()
-	pi := &prefixIterator{
-		itr:    itr,
-		prefix: prefix,
-	}
-	pi.Seek(prefix)
-	return pi
+func IsKeyInDomain(key string, start, end []byte) bool {
```

Comment:
Could we change this signature to take key in as a []byte. Notice that in every call here we are incuring
O(2*n) []byte<-->string conversions whereas by making the signature take in `key []byte`, we don't have to do any strings.Compare and string(start), string(end) comparisons. bytes already has bytes.Compare.

Reply:
yeah it should be []byte.